### PR TITLE
Introduce paging runner that simplifies scraping of paged websites

### PR DESCRIPTION
**Pull request by GermanGolota**

### CODE REVIEW

File: src/Aurora.Scrapers/Extensions/HttpClientExtensions.cs [hunk 1]
```diff
@@ -18,4 +18,25 @@ public static async Task<bool> TryLoadDocumentFromUrl(this HttpClient client, Ht
...
+            {
+                OptionFixNestedTags = true
+            };
+            var htmlContent = await response.Content.ReadAsStringAsync();
+            document.LoadHtml(htmlContent);
+            result = document;
+        }
+        else
+        {
+            result = $"Failed to load page from '{url}'".ToErrorResult<HtmlDocument>();
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Extensions/HttpClientExtensions.cs:38]: you can put it into the variable right away and omit this else statements

File: src/Aurora.Scrapers/Extensions/StringExtensions.cs [hunk 2]
```diff
@@ -1,6 +1,23 @@
-﻿namespace Aurora.Scrapers.Extensions;
+﻿using System.Text.RegularExpressions;
+
+namespace Aurora.Scrapers.Extensions;
 
 internal static class StringExtensions
 {
     public static string FormatTermToUrl(this string term) => term.Replace(" ", "+");
+
+    private static Regex _numbersRegex = new Regex("\\d+(\\.\\d+)?", RegexOptions.Multiline | RegexOptions.Compiled);
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Extensions/StringExtensions.cs:9]: did u say regex bad?

File: src/Aurora.Scrapers/Option/PornhubGifsScraper.cs [hunk 3]
```diff
@@ -1,80 +1,78 @@
...
-        for (var i = 0; i < config.MaxPagesCount; i++)
-        {
-            if (config.UseLimitations && urlsCount >= config.MaxItemsCount)
+        return await _runner.RunPagingAsync(HttpClientNames.DefaultClient,
+            loadPage: async (pageNumber, client) => await TryLoadPage(terms, baseUrl, client, pageNumber),
+            scrapPage: document =>
             {
-                break;
-            }
+                var bodyNode = document.DocumentNode?.SelectSingleNode("//body");
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Option/PornhubGifsScraper.cs:26]: check body nullability

File: src/Aurora.Scrapers/Services/PagingRunner.cs [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+    /// <param name="loadPage">Function that loads a page based on its number. Numbers start from 0</param>
+    /// <param name="scrapPage">Function that finds all of the search results given the document</param>
+    /// <param name="findMaxPageNumber">
+    /// Optional. Function that determines max possible number of pages for current search options.
+    /// Number of pages is inclusive, so if you reply that 5 pages is the max number of pages then you <paramref name="loadPage"/> would get called 5 times.
+    /// Default is being provided from configuration
+    /// </param>
+    /// <param name="pagesWaitTime">Optional. Time to wait in-between scraping pages. Default is 1/4 of a second</param>
+    /// <param name="scraperName">Would be set by automatically by the compiler, so please do not set it yourself.</param>
+    public async Task<List<SearchItem<T>>> RunPagingAsync<T>(string clientName,
```
Comment (RuslanPr0g) [inline src/Aurora.Scrapers/Services/PagingRunner.cs:32]: пизда

Comment (RuslanPr0g) [reply src/Aurora.Scrapers/Services/PagingRunner.cs:32]: гамняшка

**Review by RuslanPr0g — APPROVED** [review]

**Review by RuslanPr0g — APPROVED** [review]

File: src/Aurora.Scrapers/Services/PagingRunner.cs (hunk 4 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/Services/PagingRunner.cs:32]: true

File: src/Aurora.Scrapers/Option/PornhubGifsScraper.cs (hunk 3 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/Option/PornhubGifsScraper.cs:26]: We already do that on the next line of code:
 var gifLinksNodes = bodyNode?.SelectNodes("//li[contains(@class, 'gifVideoBlock')]");
 if (gifLinksNodes is not null)

File: src/Aurora.Scrapers/Extensions/StringExtensions.cs (hunk 2 above)
Comment (GermanGolota) [reply src/Aurora.Scrapers/Extensions/StringExtensions.cs:9]: If we ever are interested in improving performance, then we might look into optimizing that, but it is good enough with this regex

**Pull request merged**